File: web/src/routes/Game/GameRenderer.ts
import { Vector } from "protocol/dist/classes/Game";
import type { BoardConfiguration, GameConfiguration, PelletType, Pixel, Player, PlayerPositioning, Vector as IVector } from "protocol/dist/interfaces/Game";
import { Camera } from "./Camera";
import { ColorSquare, Texture } from "./Texture";

export interface GameTextureSet {
    solid: Texture,
    pellet: (pType: PelletType) => Texture,
    player: (player: Player, positioning: Pixel[], positionIndex: number) => Texture
}

const mul = (v: IVector, n: number) => new Vector(v.x * n, v.y * n);

class ColorSquareTextures implements GameTextureSet {
    public solid: ColorSquare;
    constructor(private size: number) {
        this.solid = ColorSquare.colorSquare('gray', size)
    }
    public pellet({ color }: PelletType) { return ColorSquare.colorSquare(color, this.size) }
    public player({ color }: Player) { return ColorSquare.colorSquare(color, this.size) }
}

export class GameRenderer {
    private cam: Camera;
    private ctx: CanvasRenderingContext2D;
    private textures: GameTextureSet;
    private boardConfig!: BoardConfiguration; // 'not undefined' assertion here because: https://github.com/microsoft/TypeScript/issues/36931

    public static readonly DEFAULT_BLOCK_SIZE_IN_PIXEL = 20;
    public static readonly DEFAULT_FOCUS_RATIO = 1/3;

    constructor(
        private canvas: HTMLCanvasElement,
        private canvasSize: Vector,
        private gameConfig: GameConfiguration,
        private myPlayerId: string,
        private blockSizeInPixel: number = GameRenderer.DEFAULT_BLOCK_SIZE_IN_PIXEL,
        private focusRatio: number = GameRenderer.DEFAULT_FOCUS_RATIO,
    ) {
        const ctx = canvas.getContext('2d');
        if (!ctx) throw new Error('couldn\'t get canvas context')
        this.ctx = ctx;
            
        this.textures = new ColorSquareTextures(this.blockSizeInPixel),
        this.cam = new Camera(new Vector(canvasSize.x, canvasSize.y), ctx);
    }

    updateGameConfiguration(g: GameConfiguration) {
        this.gameConfig = g;
    }

    updateBoardConfiguration(b: BoardConfiguration) {
        this.boardConfig = b;
    }

    updateCanvas(canvasSize: Vector, canvas?: HTMLCanvasElement) {
        this.canvas = canvas ?? this.canvas;
        this.canvasSize = canvasSize;

        this.cam = new Camera(new Vector(this.canvasSize.x, this.canvasSize.y), this.ctx);
    }

    canRender() {
        return this.boardConfig && this.gameConfig && this.canvas && this.canvasSize;
    }

    render = () => {
        if (!this.canRender()) throw new Error('can\'t render, you are doing something wrong')

        const myPlayer = this.getPlayerByID(this.myPlayerId);
        const focus = this.isActivePlayer(myPlayer.name)
            ? this.boardConfig.players.find(({ name }) => name === myPlayer.name)!.vectors[0]
            : new Vector(0, 0)
        console.log(focus, myPlayer)
        this.setupCamera(mul(focus, this.blockSizeInPixel))

        for (const pellet of this.boardConfig.pellets) {
            this.textures.pellet(pellet.type).draw(mul(pellet.vector, this.blockSizeInPixel), this.ctx)
        }

        for (const player of this.boardConfig.players) {
            this.drawPlayer(player);
        }

        console.log(this.gameConfig.solid)

        for (const v of this.gameConfig.solid) {
            this.textures.solid.draw(mul(v, this.blockSizeInPixel), this.ctx);
        }
    }

    private drawPlayer(playerPos: PlayerPositioning) {
        const pl = this.getPlayer(playerPos.name);
        if (!pl) throw new Error('can\'t draw non-existent player')
        for (let i = 0; i < playerPos.vectors.length; i++) {
            this.textures.player(pl, playerPos.vectors, i).draw(mul(playerPos.vectors[i], this.blockSizeInPixel), this.ctx);
        }
    }

    private isActivePlayer(oName: string): boolean {
        return !!this.boardConfig.players.find(({ name }) => oName === name);
    }

    private getPlayer(oName: string): Player {
        const player = this.gameConfig.players.find(({ name }) => oName === name)
        if (!player) throw new Error('no player with name found')
        return player;
    }

    private getPlayerByID(oId: string): Player {
        const player = this.gameConfig.players.find(({ id }) => oId === id)
        if (!player) throw new Error('no player with id found')
        return player;
    }

    private setupCamera(focus: IVector) {
        // if game fits within bounds, don't move camera at all, otherwise move camera.
        // camera position is set to players current location when the difference between active cam loc and player gets big
        // don't forget, you are moving the world, not the player, so the vector is inverted
        this.cam.init()
        this.cam.keepPointWithinAreaOfCameraWhileRespectingContextBoundaries(
            new Vector(focus.x, focus.y), new Vector(this.canvasSize.x * this.focusRatio, this.canvasSize.y * this.focusRatio), new Vector(0, 0), this.gameSize()
        )
    }

    private gameSize() {
        if (!this.gameConfig) throw new Error('no game config')
        return new Vector(this.blockSizeInPixel * this.gameConfig.size.x, this.blockSizeInPixel * this.gameConfig.size.y)
    }
}

File: mobile/lib/routes/game/game_renderer.dart
import 'package:flutter/material.dart';
import 'package:mobile/external/protocol/interfaces/game.dart';
import 'package:mobile/routes/game/camera.dart';
import 'package:mobile/routes/game/texture.dart';

Vector multiply(Vector v, int n) => Vector(x: (v.x * n), y: (v.y * n));
Offset toOffset(Vector v) => Offset(v.x.toDouble(), v.y.toDouble());

class GameRenderer extends CustomPainter { // basically the render function, but we need to update it
  final GameTextureSet textures;
  final BoardConfiguration boardConfig;
  final GameConfiguration gameConfig;

  static const defaultBlockSizeInPixel = 18;
  static const defaultFocusRatio = 1/3;
  static const defaultBackGroundColor = Color(0xFF9bbc0f);

  final String myPlayerId;
  final int blockSizeInPixel;
  final double focusRatio;
  final Paint _backgroundPaint;

  GameRenderer({
    required this.boardConfig,
    required this.gameConfig,

    required this.myPlayerId,
    this.blockSizeInPixel = defaultBlockSizeInPixel,
    this.focusRatio = defaultFocusRatio,
    backgroundColor = defaultBackGroundColor,
  }):
    _backgroundPaint = Paint()..color = backgroundColor,
    textures = ColorSquareTextures(blockSizeInPixel);

  @override
  void paint(Canvas canvas, Size size) {
    // canvas.drawLine(const Offset(0, 0), Offset(size.width, size.height), Paint()..color = Colors.red);
    // canvas.drawCircle(
    //   Offset(size.width / 2, size.height / 2),
    //   value * size.shortestSide,
    //   Paint()..color = Colors.blue,
    // );
    final myPlayer = getPlayerByID(myPlayerId);
    final focus = isActivePlayer(myPlayer.name)
      ? boardConfig.players.where((pl) => pl.name == myPlayer.name).first.vectors[0]
      : Vector(x: 0, y: 0);
    setupCamera(toOffset(mul(focus)), canvas, size); // not imp yet

    canvas.drawRect(
      Rect.fromLTWH(0, 0, gameSize().dx, gameSize().dy),
      _backgroundPaint
    );

    for (final pellet in boardConfig.pellets) {
      textures.pellet(pellet.type).draw(mul(pellet.vector), canvas);
    }

    for (final player in boardConfig.players) {
      drawPlayer(player, canvas);
    }

    for (final v in gameConfig.solid) {
      textures.solid.draw(mul(v), canvas);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;

  Player getPlayerByID(String pid) {
    final result = gameConfig.players.where((pl) => pl.id == pid);
    if (result.isEmpty) throw Exception('no player with id found');
    return result.first;
  }

  Player getPlayerByName(String pname) {
    final result = gameConfig.players.where((pl) => pl.name == pname);
    if (result.isEmpty) throw Exception('no player with name found');
    return result.first;
  }

  bool isActivePlayer(String name) {
    return boardConfig.players.any((p) => p.name == name);
  }
  
  Vector mul(Vector v) => multiply(v, blockSizeInPixel);

  void setupCamera(Offset focus, Canvas canvas, Size size) {
    final cam = Camera(canvas, size, gameSize(), Offset.zero);
    cam.centered(focus);
    cam.apply();
  }

  void drawPlayer(PlayerPositioning positioning, Canvas canvas) {
    final pl = getPlayerByName(positioning.name);
    for (var i = 0; i < positioning.vectors.length; i++) {
      textures.player(
        player: pl,
        positioning: positioning.vectors,
        positionIndex: i
      ).draw(mul(positioning.vectors[i]), canvas);
    }
  }

  Offset gameSize() => toOffset(mul(gameConfig.size));
}